### PR TITLE
Add in preliminary PyPy support

### LINKED ISSUE #5054: Workers running under PyPy attempt to load psycopg2 (not psycopg2cffi) and crash
### Description

When attempting to run Synapse 0.99.3 (with PostgreSQL and workers) under PyPy, it somehow still tries to load regular psycopg2 (instead of psycopg2cffi) and crashes.

### Steps to reproduce

- create a PyPy virtualenv
- install synapse, psycopg2cffi via pip
- start any worker

### Logs

```
Apr 13 17:53:14 stratofortress systemd[1]: Started Synapse Matrix homeserver worker federation_reader-0.
Apr 13 17:53:23 stratofortress root[23166]: [None] ***** STARTING SERVER *****
Apr 13 17:53:23 stratofortress root[23166]: [None] Server /home/operator/venv/pypy/site-packages/synapse/app/federation_reader.py version 0.99.3 (b=master,7963d79c3)
Apr 13 17:53:23 stratofortress root[23166]: [None] Server hostname: intelfx.name
Apr 13 17:53:23 stratofortress twisted[23166]: [None] Redirected stdout/stderr to logs
Apr 13 17:53:23 stratofortress synapse.server[23166]: [None] Setting up.
Apr 13 17:53:23 stratofortress synapse.config.appservice[23166]: [None] Loaded application service: ApplicationService: {'token': '<redacted>', 'url': 'http://localhost:8409', 'hs_token': '<redacted>', 'sender': '@appservice-irc:intelfx.name', 'server_name': 'intelfx.name', 'namespaces': {'users': [{'exclusive': True, 'regex': re.compile('@fn_.*:.*')}, {'exclusive': True, 'regex': re.compile('@irc_.*_.*:.*')}], 'aliases': [{'exclusive': True, 'regex': re.compile('#fn_.*:.*')}, {'exclusive': True, 'regex': re.compile('#irc_.*_.*:.*')}], 'rooms': []}, 'id': '134313c1983d17ed2560eeff6ef95db173c5e42861f82d8f07f67054a7fd458d', 'ip_range_whitelist': None, 'protocols': set(), 'rate_limited': True}
Apr 13 17:53:23 stratofortress twisted[23166]: [] Traceback (most recent call last):
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/opt/pypy3/lib-python/3/runpy.py", line 193, in _run_module_as_main
Apr 13 17:53:23 stratofortress twisted[23166]: []     "__main__", mod_spec)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/opt/pypy3/lib-python/3/runpy.py", line 85, in _run_code
Apr 13 17:53:23 stratofortress twisted[23166]: []     exec(code, run_globals)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/app/federation_reader.py", line 186, in <module>
Apr 13 17:53:23 stratofortress twisted[23166]: []     start(sys.argv[1:])
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/app/federation_reader.py", line 178, in start
Apr 13 17:53:23 stratofortress twisted[23166]: []     ss.setup()
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/server.py", line 224, in setup
Apr 13 17:53:23 stratofortress twisted[23166]: []     self.datastore = self.DATASTORE_CLASS(conn, self)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/replication/slave/storage/account_data.py", line 30, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(SlavedAccountDataStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/account_data.py", line 47, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(AccountDataWorkerStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/replication/slave/storage/pushers.py", line 26, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(SlavedPusherStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/replication/slave/storage/push_rule.py", line 28, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(SlavedPushRuleStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/replication/slave/storage/receipts.py", line 40, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(SlavedReceiptsStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/replication/slave/storage/events.py", line 61, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(SlavedEventStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/push_rule.py", line 74, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(PushRulesWorkerStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/appservice.py", line 59, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(ApplicationServiceWorkerStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/receipts.py", line 43, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(ReceiptsWorkerStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/event_push_actions.py", line 70, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(EventPushActionsWorkerStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/stream.py", line 164, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(StreamWorkerStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/state.py", line 375, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(StateGroupWorkerStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/registration.py", line 32, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(RegistrationWorkerStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/replication/slave/storage/room.py", line 24, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(RoomStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/transactions.py", line 61, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(TransactionStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/replication/slave/storage/_base.py", line 37, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     super(BaseSlavedStore, self).__init__(db_conn, hs)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/storage/_base.py", line 203, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     self._db_pool = hs.get_db_pool()
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/server.py", line 540, in _get
Apr 13 17:53:23 stratofortress twisted[23166]: []     dep = builder()
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/synapse/server.py", line 404, in build_db_pool
Apr 13 17:53:23 stratofortress twisted[23166]: []     **self.db_config.get("args", {})
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/twisted/enterprise/adbapi.py", line 211, in __init__
Apr 13 17:53:23 stratofortress twisted[23166]: []     self.dbapi = reflect.namedModule(dbapiName)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/twisted/python/reflect.py", line 159, in namedModule
Apr 13 17:53:23 stratofortress twisted[23166]: []     topLevel = __import__(name)
Apr 13 17:53:23 stratofortress twisted[23166]: []   File "/home/operator/venv/pypy/site-packages/psycopg2/__init__.py", line 50, in <module>
Apr 13 17:53:23 stratofortress twisted[23166]: []     from psycopg2._psycopg import (                     # noqa
Apr 13 17:53:23 stratofortress twisted[23166]: [] ImportError: /home/operator/venv/pypy/site-packages/psycopg2/_psycopg.pypy3-71-x86_64-linux-gnu.so: undefined symbol: timeradd
Apr 13 17:53:23 stratofortress systemd[1]: synapse-worker@federation_reader-0.service: Main process exited, code=exited, status=1/FAILURE
Apr 13 17:53:23 stratofortress systemd[1]: synapse-worker@federation_reader-0.service: Failed with result 'exit-code'.
```

### Version information

- **Homeserver**: intelfx.name
- **Version**: Synapse 0.99.3 / PyPy 7.1.0
- **Install method**: pip
- **Platform**: Arch GNU/Linux

### PR DESCRIPTION
**Pull request by ShadowJonathan**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Related to #8888
Fixes #5054

`Signed-off-by: Jonathan de Jong <jonathan@automatia.nl>`

### CODE REVIEW

**Comment by ShadowJonathan** [comment]
Blocked by https://github.com/pallets/markupsafe/issues/176 (Unicode C API messing up), because this affects `jinja2.escape`, which is pretty vital to jinja templating, and so I wouldn't continue on with this if i wasn't sure this affected other areas of the server.

After they release version 2 (with hopefully this fixed), and `jinja2` is verified to work with v2, I think adding `"markupsafe>=2 ; platform_python_implementation != 'PyPy'"` to dependencies will fix that issue.

**Comment by mattip** [comment]
Can I see the failing log for PyPy? I want to exactly replicate the environment to debug the PyPy c-api call failure.

**Comment by ShadowJonathan** [comment]
@mattip it's local, but the command is `pypy3 -m tox -re pypy3 -- tests.rest.client.v1.test_login.MultiSSOTestCase.test_multi_sso_redirect`, after installing `tox` on `pypy3`(.6), to invoke that failing test.

**Comment by mattip** [comment]
It does not reproduce for me. What version of PyPy are you using?

**Comment by ShadowJonathan** [comment]
@mattip 
```
jonathan@jon:~$ pypy3 --version
Python 3.6.9 (7.3.1+dfsg-4ubuntu1, Sep 17 2020, 15:27:03)
[PyPy 7.3.1 with GCC 10.2.0]
```

**Comment by mattip** [comment]
Ahh. Yes, there are unicode c-api bugs in that version. I guess that is the one available via `apt get`? The current version (7.3.3) would be the one you want to use. I see you are using a docker on circleCI for CI? You might want to use the docker images that cibuildwheel provides, they have up-to-date versions of python including pypy. Or you could move to github actions, they also provide up-to-date version of pypy. If you want to do the minimum change, you can use the release tarballs from https://downloads.python.org/pypy/.  They are portable and should work on any machine that supports manylinux2010.
```
pushd /tmp
wget https://downloads.python.org/pypy/pypy3.6-v7.3.3-linux64.tar.bz2
tar -xf pypy3.6-v7.3.3-linux64.tar.bz2
export PATH=/tmp/pypy3.6-v7.3.3-linux64.tar/bin;$PATH
ln -s /tmp/pypy3.6-v7.3.3-linux64.tar/bin/python /tmp/pypy3.6-v7.3.3-linux64.tar/bin/pypy3
python -m ensurepip
python -m install tox 
```
or if you prefer conda
```
conda create -n pypy36 pypy3.6
conda activate pypy36
conda install -y tox
```

**Comment by ShadowJonathan** [comment]
Docker image `pypy:3.6` is PyPy 7.3.3, I think we'll be alright there.

Thanks for the headsup about the unicode bugs in that pypy version. :+1:

> I see you are using a docker on circleCI for CI?

Docker on buildkite, with boxes provided by matrix.org themselves, see the [pipeline](https://github.com/matrix-org/pipelines/blob/master/synapse/pipeline.yml) for more details.

**Comment by ShadowJonathan** [comment]
Some test results:

`-j 2`
`py36`: `Ran 1501 tests in 875.411s`
`pypy3`: `Ran 1501 tests in 4378.433s`

`-j 3`
`py36-postgres`: `Ran 1501 tests in 706.022s`
`pypy3-postgres`: `Ran 1501 tests in 2548.308s`

:neutral_face:

**Comment by ShadowJonathan** [comment]
https://www.pypy.org/performance.html tells that (perfect unit) tests and short-ran scripts don't get time to be optimised by the JIT

So tests are severely affected by the JIT this way, i'm going to look for ways to front-load this problem, and instead "warm up" the JIT to make it possible for tests to run more smoothly.

**Comment by ShadowJonathan** [comment]
Looking further into it, i get the distinct impression pypy might have a C-API performance issue, [this gist](https://gist.github.com/wizzat/7112622#file-psycopg2_performance-py-L111-L135) shows the performance difference between python and pypy on psycopg2.

I found [pg8000](https://pypi.org/project/pg8000/) which is written in pure python, this might be counter-intuitive to use, but it's possible that with a JIT-optimized pypy, this library can outperform psycopg2, as with the tests, i saw them being done at blazing speeds when i removed a stray `gc.collect()` call somewhere, so this might be worth looking into.

**Comment by mattip** [comment]
> pypy might have a C-API performance issue

Definitely, emulating the C-API is slow. There is work ongoing to improve this via a new c-api ([HPy](https://hpy.readthedocs.io/en/latest/)) that will layer this on top of the current c-api. 

Writing simple python code (as little dynamic class creation as possible, keeping class heirarchies shallow) will be fast on PyPy, for interfacing with C we recommend CFFI and for C++ we recommend cppyy.

**Comment by ShadowJonathan** [comment]
Looking at https://morepypy.blogspot.com/2018/09/inside-cpyext-why-emulating-cpython-c.html, i got the idea to use [pg8000](https://pypi.org/project/pg8000/), a pure-python DB API v2-complicit postgres driver. I'm currently running the tests again (at `-j 3`) using this driver, so we'll see how good the performance is, the early results (test OKs breezing past my screen) are promising.

**Comment by ShadowJonathan** [comment]
Just finished running the new `pypy3-postgres` suite with `-j 3` and `pg8000` as driver: `Ran 1501 tests in 763.756s` :D

**Comment by ShadowJonathan** [comment]
To reiterate: This PR will only add in some simple support for PyPy. In it's current form, a running server would be *abysmally* slow, 

`pg8000` will be added in another PR, and additional DB-cleaning-up PRs after that (that make adding extra adapters/drivers/engines easier, especially with the prospect of PEP 249 existing, and will seek to clean up the codebase to start to use `with engine.new_connection() as conn:` context managers and such). I'll also think about making a `"postgres"` virtual adapter then, for reverse compatibility, all `"psycopg2"`-specified driver configurations will keep using psycopg2, and wont suddenly switch over to `pg8000` then, with `"postgres"` as "engine identifier", it'll switch between `psycopg2` and `pg8000` depending if the current running python is pypy or python. I think fancy things such as replicated writers will have to wait in pypy, as i first want to vet `pg8000` before that happens.

I'll add the `-prof` `cProfile`ing tox environment in another PR.

**Comment by callahad** [comment]
> In it's current form, a running server would be abysmally slow, due to C-API interaction with it's internal sqlite3 package, and C-API interaction with psycopg2cffi.

I believe you're misunderstanding; psycopg2cffi, per the name, uses CFFI, as mattip recommends. This is in contrast to CPython's native C extension API documented at https://docs.python.org/3/extending/extending.html.

**Comment by ShadowJonathan** [comment]
It's still slow, the above test results (`pypy3-postgres: Ran 1501 tests in 2548.308s`) are using `psycopg2cffi`, so even if cffi is supposed to give a performance boost (or mitigate pypy's C API translation), I can't find evidence of it.

**Review by clokep — COMMENTED** [review]
Looks pretty reasonable overall! To be clear -- with these changes tests will pass? Did you run sytests by any chance?

**Comment by mattip** [comment]
Thanks for doing this. It would be interesting to see how this plays out in a production environment where the JIT can actually make things faster.

When thinking about which version of PyPy to test (assuming you will not test both python3.6 and python3.7), I would suggest you go with 3.7 not 3.6. Going forward, people are starting to drop 3.6 since it is no longer supported, and when underlying dependencies are updated they may not be available for 3.6.

File: synapse/storage/engines/__init__.py
Comment (ShadowJonathan) [reply synapse/storage/engines/__init__.py:30]: It is here in case there is confusion with a "wait, `psycopg2` is installed?" question whenever the `python_dependencies.py` is examined and cross-referenced with this, an assurance that "`psycopg2`" is able to be properly imported on pypy after all.

File: synapse/storage/engines/sqlite.py
Comment (ShadowJonathan) [reply synapse/storage/engines/sqlite.py:37]: Back when i made this change, there were still a few of them left, and so I made this change to assure they were caught anyway.

I can remove this, but i'd rather not, because i think it is pretty possible `bytearray`s are re-introduced somewhere else, which then gets caught on this.

File: tests/__init__.py
Comment (ShadowJonathan) [reply tests/__init__.py:24]: Should i revert this?

File: synapse/storage/engines/__init__.py
Comment (clokep) [reply synapse/storage/engines/__init__.py:30]:
```suggestion
        # Note that psycopg2cffi-compat provides the psycopg2 module on mypy.
```

File: synapse/python_dependencies.py
Comment (ShadowJonathan) [reply synapse/python_dependencies.py:92]: That comment says it should be *compatible*, not necessary, we bumped to `2.8` before this, and `psycopg2cffi` already supported `2.8`

**Comment by ShadowJonathan** [comment]
> To be clear -- with these changes tests will pass? Did you run sytests by any chance?

I ran `tox -e pypy3` tests, but i haven't yet ran `sytest` tests, i'll do that in a second.

File: tests/__init__.py
Comment (clokep) [reply tests/__init__.py:24]: I think for now, yes.

File: synapse/storage/engines/sqlite.py
Comment (clokep) [inline synapse/storage/engines/sqlite.py:36]:
```suggestion
            # pypy's sqlite3 module doesn't handle bytearrays, convert them
            # back to bytes.
```

File: synapse/storage/engines/sqlite.py
Comment (clokep) [reply synapse/storage/engines/sqlite.py:37]: Fair enough.

File: changelog.d/9123.misc
Comment (clokep) [inline changelog.d/9123.misc:1]:
```suggestion
Add experimental support for running Synapse with PyPy.
```

File: tests/__init__.py
Comment (clokep) [inline tests/__init__.py:24]:
```suggestion
util.DEFAULT_TIMEOUT_DURATION = 20
```

File: tests/unittest.py
Comment (clokep) [inline tests/unittest.py:131]: Does this call break or is it just slow? I'm a bit nervous that this will cause tests to not fail in some situations when they should?

File: tests/__init__.py
Comment (ShadowJonathan) [reply tests/__init__.py:24]: The variable should be a float, as per the documentation: https://twistedmatrix.com/documents/current/api/twisted.trial.util.html

File: tests/unittest.py
Comment (ShadowJonathan) [reply tests/unittest.py:131]: "to workaround problems with deferreds leaking logcontexts when they are GCed"

I recently heard a new twisted version will fix this problem, per @richvdh, but i'm not sure.

In the time i've tested this so far, nothing has broken, but i think we need a second opinion on this one, then.

Comment (clokep) [reply tests/unittest.py:131]: I guess my question was more of "what happens if you call `gc.collect()` on pypy?"

File: tests/__init__.py
Comment (clokep) [reply tests/__init__.py:24]: It is an unrelated change in this PR, please remove it.

File: tests/unittest.py
Comment (ShadowJonathan) [reply tests/unittest.py:131]: It's fairly expensive, pypy has it's own GC implementation and the like, and looking at profile runs, it showed up fairly high (thus was pretty slow)

Comment (clokep) [reply tests/unittest.py:131]: Can we add that to the comment so that in the future we remember why we're skipping it?

**Review by clokep — APPROVED** [review]

**Pull request merged**
